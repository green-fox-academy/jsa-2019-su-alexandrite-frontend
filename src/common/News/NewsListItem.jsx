import React from 'react';
import {
  Text,
  Image,
  ViewPropTypes,
} from 'react-native';
import PropTypes from 'prop-types';
import { TouchableHighlight } from 'react-native-gesture-handler';
import { useNavigation } from 'react-navigation-hooks';
import Column from '../Column';
import commonStyles from '../styles';

const NewsListItem = ({ item, style }) => {
  const date = new Date(item.datetime);
  const { push } = useNavigation();
  return (
    <TouchableHighlight
      onPress={() => push('NewsWebView', { url: item.url })}
      underlayColor="#eee"
      style={[style, commonStyles.newsItem]}
    >
      <>
        <Image
          source={item.image ? { uri: item.image } : null}
          style={commonStyles.newsImg}
        />
        <Column flex={1} style={{ justifyContent: 'space-between' }}>
          <Text style={commonStyles.newsHeadline} lineBreakMode="tail" numberOfLines={3}>{item.headline}</Text>
          <Text style={commonStyles.newsTime}>{date.toLocaleString()}</Text>
        </Column>
      </>
    </TouchableHighlight>
  );
};

NewsListItem.propTypes = {
  item: PropTypes.shape({
    datetime: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    headline: PropTypes.string,
    url: PropTypes.string,
    image: PropTypes.string,
  }).isRequired,
  style: ViewPropTypes.style,
};

NewsListItem.defaultProps = {
  style: undefined,
};

export default NewsListItem;
