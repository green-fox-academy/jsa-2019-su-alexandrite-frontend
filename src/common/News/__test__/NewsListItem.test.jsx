import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { useNavigation } from 'react-navigation-hooks';
import MockDate from 'mockdate';

import NewsListItem from '../NewsListItem';

jest.mock('react-navigation-hooks');
jest.mock('../useFetchStockNews');

MockDate.set('2020-01-10T04:46:05.638Z');

const sampleNewsItem = {
  datetime: '2020-01-10T04:46:05.638Z',
  headline: 'headline',
  url: 'http://test.test',
  image: 'http://test.test/test.jpg',
};

describe('<NewsListItem />', () => {
  beforeEach(() => {
    useNavigation.mockReturnValue({
      push: jest.fn(),
      navigate: jest.fn(),
    });
  });

  afterAll(() => {
    jest.resetModules();
  });

  it('Should render properly and match the snapshot', () => {
    const tree = renderer.create(
      <NewsListItem item={sampleNewsItem} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });


  it('Should render properly and match the snapshot when image url not present', () => {
    const tree = renderer.create(
      <NewsListItem item={{ ...sampleNewsItem, image: null }} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });


  it('Should navigate to NewsWebView with url passed', () => {
    const mockPush = jest.fn();
    useNavigation.mockReturnValueOnce({
      push: mockPush,
    });
    const wrapper = shallow(
      <NewsListItem item={sampleNewsItem} />,
    );
    wrapper.find('TouchableHighlight').first().simulate('press');
    expect(mockPush).toHaveBeenCalledWith('NewsWebView', { url: sampleNewsItem.url });
  });
});
