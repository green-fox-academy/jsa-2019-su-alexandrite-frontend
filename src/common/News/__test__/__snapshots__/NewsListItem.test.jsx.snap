// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsListItem /> Should render properly and match the snapshot 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleColor={0}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    style={
      Object {
        "borderRadius": 6,
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Object {
          "backgroundColor": "#ccc",
          "borderRadius": 6,
          "height": 169,
          "overflow": "hidden",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "http://test.test/test.jpg",
          }
        }
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": 169,
              "width": undefined,
            },
            undefined,
          ]
        }
      />
      <ViewManagerAdapter_ExpoLinearGradient
        proxiedProperties={
          Object {
            "colors": Array [
              0,
              0,
              2281701376,
              4278190080,
            ],
            "endPoint": undefined,
            "locations": undefined,
            "startPoint": undefined,
          }
        }
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
            "padding": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 12,
            }
          }
        >
          1/1/1970, 8:00:00 AM
        </Text>
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 14,
              "fontWeight": "bold",
              "lineHeight": 20,
            }
          }
        >
          headline
        </Text>
      </ViewManagerAdapter_ExpoLinearGradient>
    </View>
  </View>
</RNGestureHandlerButton>
`;

exports[`<NewsListItem /> Should render properly and match the snapshot when image url not present 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleColor={0}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    style={
      Object {
        "borderRadius": 6,
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Object {
          "backgroundColor": "#ccc",
          "borderRadius": 6,
          "height": 169,
          "overflow": "hidden",
        }
      }
    >
      <Image
        source={null}
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": 169,
              "width": undefined,
            },
            undefined,
          ]
        }
      />
      <ViewManagerAdapter_ExpoLinearGradient
        proxiedProperties={
          Object {
            "colors": Array [
              0,
              0,
              2281701376,
              4278190080,
            ],
            "endPoint": undefined,
            "locations": undefined,
            "startPoint": undefined,
          }
        }
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
            "padding": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 12,
            }
          }
        >
          1/1/1970, 8:00:00 AM
        </Text>
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 14,
              "fontWeight": "bold",
              "lineHeight": 20,
            }
          }
        >
          headline
        </Text>
      </ViewManagerAdapter_ExpoLinearGradient>
    </View>
  </View>
</RNGestureHandlerButton>
`;
